Added the absolute state of KDE software that should get released in lockstep with the KDE megarelease (#9)
Greybeard-Entertainment/pages: docs/rust-is-oop.org

#+AUTHOR: Aleksandr Petrosyan
#+TITLE: Rust is an OOP langauge

And a terrible functional one.  As in yes, there are anonymous functions (Java has those too), there are typeclasses, (if you squint hard enough, it's just like interfaces in Java), and in fact, pattern matching with tagged unions as a first class citizen can be the only thing that can even with some gusto be argued to be a functional feature.

But I'm not here to pounce on Rust's uselessness in teaching functional programming, that is a topic for another article.  I don't hate Rust and I don't want to fling more negativity its way than I have to.

> But I will argue that Rust is truer to the original vision of OOP than the half-hearted attempts in Python, and the principled, though baroque approach adopted in Java.  It is not quiet on the same level as SmallTalk, JavaScript or Objective C, but those languages are largely misunderstood.

Reviewer: ```suggestion
But I will argue that Rust is truer to the original vision of OOP than the half-hearted attempts in Python, and the principled, though baroque approach adopted in Java.  It is not quite on the same level as SmallTalk, JavaScript or Objective C, but those languages are largely misunderstood.
```

> But I will argue that Rust is truer to the original vision of OOP than the half-hearted attempts in Python, and the principled, though baroque approach adopted in Java.  It is not quiet on the same level as SmallTalk, JavaScript or Objective C, but those languages are largely misunderstood.

Reviewer: ```suggestion
I will argue that Rust is truer to the original vision of OOP than the half-hearted attempts in Python, and the principled, though baroque approach adopted in Java.  It is not quiet on the same level as SmallTalk, JavaScript or Objective C, but those languages are largely misunderstood.
```

Let's illustrate on an example, perhaps a bit more contrived.  In an OOP world, there is no state, there are only signals.  The only way to communicate with an object to query its state is via signals.  But most signals are trivial.   A programming language like SmallTalk solved this problem easily, and the modern day problems related to it have more to do with the refusal to let go of the procedural aspects of programming than with the ineptitude(s) of the OOP paradigm.  Case in point, if one wanted to model the animal kingdom[fn:4], there are various signals one can process, like =eat=, =drink=, =reproduce= which are not simple queries of one's state[fn:3].  Yet, they are all trivial.  In the domain of the problem, most diets fall into categories of carnivore and herbivore, reproduction is similarly categorised very simply, with some notable outliers.  *Most of the signals for most of the objects are going to be identical.*

Now here I will make a distinction in how OOP languages traditionally solved the problem of repeated code, and why the solution in Rust is better, though not fundamentally different.

> The first key assumption is that there is a clear hierarchy.  Within the C world, one would make a table for all unique functions and then point each individual entity's behaviour function pointer to that function.  So adding each new species entity isn't a hard task.  But the amount of work is certainly sub-optimal.  So one needs to spend a bit more time and see that there are categories of herbivore and Carnivore, vertebrate vs invertebrate, and to come up with a tree of inheritance that tends to categorise animals by which behaviour they engage in.  Sometimes, if the behaviour follows from another behaviour, and as a crucial difference from the function pointer approach, one can make a shortcut.  Java, Rust, Haskell, and C++ all do it that way.   But the devil is in the details.

Reviewer: ```suggestion
The first key assumption is that there is a clear hierarchy.  Within the C world, one would make a table for all unique functions and then point each individual entity's behaviour function pointer to that function.  So adding each new species entity isn't a hard task.  But the amount of work is certainly sub-optimal.  So one needs to spend a bit more time and see that there are categories of herbivore and Carnivore, vertebrate vs invertebrate, and to come up with a tree of inheritance that tends to categorise animals by which behaviour they engage in.  Sometimes, if the behaviour follows from another behaviour, and as a crucial difference from the function pointer approach, one can make a shortcut.  Java, Rust, Haskell, and C++ all do it that way.   But the devil is in the details.
```
```suggestion
The first key assumption is that there is a clear hierarchy.  Within the C world, one would make a table for all unique functions and then point each individual entity's behaviour function pointer to that function.  So adding each new species entity isn't a hard task.  But the amount of work is certainly sub-optimal.  So one needs to spend a bit more time and see that there are categories of herbivore and carnivore, vertebrate vs invertebrate, and to come up with a tree of inheritance that tends to categorise animals by which behaviour they engage in.  Sometimes, if the behaviour follows from another behaviour, and as a crucial difference from the function pointer approach, one can make a shortcut.  Java, Rust, Haskell, and C++ all do it that way.   But the devil is in the details.
```

> With Java and C++, the assumption is that the participation in a particular way is tied to the object's self.  A cat, isn't just something that implements =Carnivore= but rather being a =Cat= is inextricable from being a =Carnivore=.  While it does make sense for a cat, consider what the inheritance hierarchy for a Platypus would look like.  Or any of a number of species that have convergent-evolved into a Crab.

Reviewer: - Can it be written that way or is it written like "evolved convergently"?
- Do you point out the part of classification through the capitalisation?

Author: This is the convention of Rust and Haskell to capitalise the Types of things.

With Haskell and Rust, one is given a way to make use of the categorical shortcuts, without making it too much a part of the type.  So a =Cat= happens to be =Carnivore=, being a =Carnivore= implies you have a mouth, so the ability to =drink= becomes obvious.  But at the same time, something like a =Platypus= isn't necessarily any of those things.  It can fall outside the regular categories and implement the basic low-level interfaces.

> Now it might seem that because the technical difference is subtle, that the resulting code is also different only subtly.  But that is far from the case.  In Fallout New Vegas, the vehicles are hats attached to an entity.  This is not what a non-technical person would assume does the walking, but the way the code is organised makes it harder to do anything else.  With Haskell and Rust, the act of moving is less tightly coupled, and that gives one more room to manoeuvre with semantics.  This approach is more granular, and discourages trying to model every aspect of the domain taxonomic-ally.  You're writing procedural code where it's more convenient, but if you need to impose a hierarchy of interfaces, they are there and very minimal.

Reviewer: ```suggestion
Now it might seem that because the technical difference is subtle, that the resulting code is also different only subtly.  But that is far from the case.  In “Fallout New Vegas”, the vehicles are hats attached to an entity.  This is not what a non-technical person would assume does the walking, but the way the code is organised makes it harder to do anything else.  With Haskell and Rust, the act of moving is less tightly coupled, and that gives one more room to manoeuvre with semantics.  This approach is more granular, and discourages trying to model every aspect of the domain taxonomically.  You're writing procedural code where it's more convenient, but if you need to impose a hierarchy of interfaces, they are there and very minimal.
```

So with Rust, you are actually imposing a hierarchical structure, you're just not being overly sentimental about it, and you apply taxonomy based on practical considerations inherent in category theory, and not replicating and badly designing an object model /ad hoc/.

* Encapsulation

One of the most important aspects of OOP is the establishment of boundaries.  Alan Kay's original vision is that state is only meaningful from the perspective of the inherent methods.  Naturally this leads to even more code being required to query the state outside the object, more work and more thinking.  So programming languages like Java allowed some state to be public, similarly to C, but unlike C, any attempt to access the inaccessible state is a compilation error[fn:5].

So where does that leave us?

> Well, in Rust pretty much all the same is present.  You can choose the visibility of an object's properties almost universally and granular.  This is as good as Java, because I can control the access to any field, without having to change its place.  Moreover, unlike having to remember which vague statement corresponds to which scope, I have a simple =crate= meaning "translation unit", =super= meaning "also in the enclosing module", and plain, meaning "in general".  This makes it more likely that I won't frivolously add a "getter" and a "setter" which essentially do the exact same as public access would.

Reviewer: ```suggestion
Well, in Rust pretty much all the same is present.  You can choose the visibility of an object's properties almost universally, in a granular fashion.  This is as good as Java, because I can control the access to any field, without having to change its place.  Moreover, unlike having to remember which vague statement corresponds to which scope, I have a simple =crate= meaning "translation unit", =super= meaning "also in the enclosing module", and plain, meaning "in general".  This makes it more likely that I won't frivolously add a "getter" and a "setter" which essentially do the exact same as public access would.
```

* Conclusion

Rust is OOP.  It is in fact, OOP at its best so far.  It is no way comparable to OCaml (and definitely not StandardML), but it is comparable to C++ or SmallTalk, even though those languages still have certain advantages, nothing in tech is ever a direct upgrade.

* Footnotes
[fn:5]: Some are under the false impression that this implies that the state *cannot* be accessed outside that object.  This is completely untrue, because knowing the layout of the object allows one to read the end-point of a raw pointer, in languages other than Java.  And if you think that you will always only be interacting with Java from Java, you are sorely mistaken.

[fn:4]: Notice how none of the neat examples in OOP, come from actual life situations and almost always are reduced to contrived examples where a clear class hierarchy of the type "is a" can be established.  This is why OOP has fallen out of favour these days.

> [fn:3]:  Outside the OOP world, in a procedural language like C, every field of every structure is always available.  So you can always do =thing.state= to read the state of an object.  In OOP, this is problematic, so in something like C++, you are encouraged to either delineate which translation units are allowed to access the state of an object and which aren't.  But in some cases, access can mean either read or write.  Instead of coming up with a permission system a-la Unix, we have instead invented a convention, wherein we restrict both read and write access, and when we need read access, we call =thing.get_state()=, or =thing.set_state(new_thing_state)= where the method =set_state= is responsible for ensuring that the object remains in a valid state[fn:2].  But here is the problem.  This led people to make "safe" assumptions about what the state can be.  So idiotic guidelines along the lines of "if it doesn't have to be publicly visible, make it private" were invented, without actually telling people that...  you know...  there aren't always resources to make the call =thing.set_state(new_thing_state)= to compile down to what the programmer actually wants.  This is a minor problem, compared to yet another guideline: "mark every member of every class that is private with a dumb prefix of your manager's choosing".  For example, if you happen to be working with Qt, your private variables are to be prefixed with =m_=, as if we can't already tell that you're a moron who refuses to use a machine that can properly differentiate private and public members in the editor window.

Reviewer: "Machine", not "software"? Does it imply some CS term?

Author: Yes, we are talking about Turing machines.
